File: cpp/test/device_build_entity_test.cpp
// Generated basic-flow test for the device_build entity (model-driven,
// unit mode; mirrors the rust/go TestEntity generator).

#include "runner_support.hpp"

using namespace sdk;
using namespace sdk::rs;

struct DeviceBuildSetup {
  std::shared_ptr<BluefinDecryptxP2peSDK> client;
  Value data;
  Value idmap;
  Value env;
  bool live = false;
  bool synthetic_only = false;
  long long now = 0;
};

static DeviceBuildSetup device_build_basic_setup(const Value& extra) {
  load_env_local();

  std::string entity_data_file = "../.sdk/test/entity/device_build/DeviceBuildTestData.json";
  Value entity_data = vs::parse_json(read_file(entity_data_file));

  Value options = vmap({{"entity", getp(entity_data, "existing")}});
  auto client = BluefinDecryptxP2peSDK::testSDK(options, extra);

  // idmap via transform (upper-cased id name synthetics), matching the donors.
  Value idmap = Struct::transform(
      vlist({Value("device_build01"), Value("device_build02"), Value("device_build03")}),
      vmap({{"`$PACK`", vlist({
        Value(""),
        vmap({
          {"`$KEY`", Value("`$COPY`")},
          {"`$VAL`", vlist({Value("`$FORMAT`"), Value("upper"), Value("`$COPY`")})}
        })
      })}}));
  if (!idmap.is_map()) idmap = vmap();

  Value env = env_override(vmap({
    {"BLUEFINDECRYPTXP_PE_TEST_DEVICE_BUILD_ENTID", idmap},
    {"BLUEFINDECRYPTXP_PE_TEST_LIVE", Value("FALSE")},
    {"BLUEFINDECRYPTXP_PE_TEST_EXPLAIN", Value("FALSE")}
  }));

  Value idmap_resolved = Helpers::toMapAny(getp(env, "BLUEFINDECRYPTXP_PE_TEST_DEVICE_BUILD_ENTID"));
  if (!idmap_resolved.is_map()) idmap_resolved = idmap;

  bool live = getp(env, "BLUEFINDECRYPTXP_PE_TEST_LIVE") == Value("TRUE");

  DeviceBuildSetup s;
  s.client = client;
  s.data = entity_data;
  s.idmap = idmap_resolved;
  s.env = env;
  s.live = live;
  s.synthetic_only = false;
  s.now = now_ms();
  return s;
}

static void device_build_entity_instance() {
  auto testsdk = BluefinDecryptxP2peSDK::testSDK();
  auto ent = testsdk->device_build();
  ASSERT_EQ(ent->getName(), std::string("device_build"), "entity name");
}

static void device_build_entity_stream() {
  // stream() runs the list op through the full pipeline and returns the
  // result items. Seed two entities via test mode; with the streaming feature
  // active it yields the feature's incremental items, else it falls back to
  // the materialised items — either way every item is yielded.
  Value seed = vmap({{"entity", vmap({{"device_build", vmap({
      {"strm01", vmap({{"id", Value("strm01")}})},
      {"strm02", vmap({{"id", Value("strm02")}})}})}})}});
  Value sdkopts = vmap({{"feature",
      vmap({{"streaming", vmap({{"active", Value(true)}})}})}});

  auto strsdk = BluefinDecryptxP2peSDK::testSDK(seed, sdkopts);
  auto se = strsdk->device_build();
  std::vector<Value> items = se->stream("list", Value::undef(), Value::undef());
  ASSERT_EQ((int)items.size(), 2, "stream yields both seeded items");

  auto plainsdk = BluefinDecryptxP2peSDK::testSDK(seed, Value::undef());
  auto pe = plainsdk->device_build();
  std::vector<Value> pitems = pe->stream("list", Value::undef(), Value::undef());
  ASSERT_EQ((int)pitems.size(), 2, "fallback stream yields both items");
}

static void device_build_entity_basic() {
  auto setup = device_build_basic_setup(Value::undef());
  std::string mode = setup.live ? "live" : "unit";
  for (const std::string& op : std::vector<std::string>{"list", "load"}) {
    auto sk = is_control_skipped("entityOp", std::string("device_build.") + op, mode);
    if (sk.first) { std::cerr << "skip: " << (sk.second.empty()? "sdk-test-control.json" : sk.second) << "\n"; return; }
  }
  auto client = setup.client;

  // Bootstrap entity data from existing test data (no create step in flow).
  // Declare _data at FUNCTION scope (later load/update steps reference it);
  // only _data_raw was declared, so the block-local assignment left _data
  // undeclared ("was not declared in this scope").
  Value device_build_ref01_data_raw = Helpers::toMapAny(Struct::getpath(setup.data, {"existing", "device_build"}));
  Value device_build_ref01_data = vmap();
  {
    std::vector<Value> its = Struct::items(device_build_ref01_data_raw);
    device_build_ref01_data = its.empty() ? vmap() : Helpers::toMapAny(pair_val(its[0]));
    if (!device_build_ref01_data.is_map()) device_build_ref01_data = vmap();
  }
  // LIST
  auto device_build_ref01_ent = client->device_build();
  Value device_build_ref01_match = vmap();
  Value device_build_ref01_list = device_build_ref01_ent->list(Struct::clone(device_build_ref01_match), Value::undef());
  ASSERT_TRUE(device_build_ref01_list.is_list(), "expected list result to be an array");

  // LOAD
  Value device_build_ref01_match_dt0 = vmap({{"id", getp(device_build_ref01_data, "id")}});
  Value device_build_ref01_data_dt0_loaded = device_build_ref01_ent->load(Struct::clone(device_build_ref01_match_dt0), Value::undef());
  Value device_build_ref01_data_dt0_load_result = Helpers::toMapAny(device_build_ref01_data_dt0_loaded);
  ASSERT_TRUE(device_build_ref01_data_dt0_load_result.is_map(), "expected load result to be a map");
  ASSERT_EQ_VAL(getp(device_build_ref01_data_dt0_load_result, "id"), getp(device_build_ref01_data, "id"), "expected load result id to match");

}

int main() {
  T_RUN(device_build_entity_instance);
  T_RUN(device_build_entity_stream);
  T_RUN(device_build_entity_basic);
  return sdktest::summary("device_build_entity_test");
}

File: scala/sdktest/DeviceBuildEntityTest.scala
// Generated basic-flow test for the device_build entity (model-driven;
// mirrors the java TestEntity generator). A dependency-free scala-cli test
// object driven by SdkEntityTestMain. Runs against the in-memory test
// transport seeded with the shipped DeviceBuildTestData.json fixtures.

import java.util.{ArrayList, LinkedHashMap, List => JList, Map => JMap}

import voxgig.bluefindecryptxp2pesdk.core.{Helpers, SdkEntity, BluefinDecryptxP2peSDK}
import voxgig.bluefindecryptxp2pesdk.utility.struct.Struct

object DeviceBuildEntityTest {

  def run(rep: SdkTestReport): Unit = {
    rep.scope("device_build.instance") {
      val testsdk = BluefinDecryptxP2peSDK.testSDK()
      val ent = testsdk.deviceBuild(null)
      rep.check("device_build.instance", ent != null, "expected non-null device_build entity")
    }

    rep.scope("device_build.basic") {
      val entityData = Helpers.toMapAny(SdkTestSupport.readJson(
          "../.sdk/test/entity/device_build/DeviceBuildTestData.json"))
      val options = new LinkedHashMap[String, Object]()
      options.put("entity", entityData.get("existing"))
      val client = BluefinDecryptxP2peSDK.testSDK(options, null)

      val idmap = new LinkedHashMap[String, Object]()
      idmap.put("device_build01", "DEVICE_BUILD01")
      idmap.put("device_build02", "DEVICE_BUILD02")
      idmap.put("device_build03", "DEVICE_BUILD03")
      val now = System.currentTimeMillis()
      val deviceBuildRef01DataRaw = Struct.items(Helpers.toMapAny(
          Struct.getpath(entityData, "existing.device_build")))
      val deviceBuildRef01Data = Helpers.toMapAny(deviceBuildRef01DataRaw.get(0).get(1))

      // LIST
      val deviceBuildRef01Ent = client.deviceBuild(null)
      val deviceBuildRef01Match = new LinkedHashMap[String, Object]()
      val deviceBuildRef01ListResult = deviceBuildRef01Ent.list(deviceBuildRef01Match, null)
      rep.check("device_build.list.islist", deviceBuildRef01ListResult.isInstanceOf[JList[?]], "expected list result to be an array, got " + deviceBuildRef01ListResult)

      // LOAD
      val deviceBuildRef01MatchDt0 = new LinkedHashMap[String, Object]()
      deviceBuildRef01MatchDt0.put("id", deviceBuildRef01Data.get("id"))
      val deviceBuildRef01DataDt0Loaded = deviceBuildRef01Ent.load(deviceBuildRef01MatchDt0, null)
      val deviceBuildRef01DataDt0LoadResult = Helpers.toMapAny(deviceBuildRef01DataDt0Loaded match { case e: SdkEntity => e.data(); case o => o })
      rep.check("device_build.load.map", deviceBuildRef01DataDt0LoadResult != null, "expected load result to be a map")
      rep.eq("device_build.load.id", deviceBuildRef01Data.get("id"), deviceBuildRef01DataDt0LoadResult.get("id"))
    }
  }
}
